src/Servers/Kestrel/Core/src/Internal/Http/Http1OutputProducer.cs

```diff
@@ -288,8 +311,56 @@ private void WriteResponseHeadersInternal(ref BufferWriter<PipeWriter> writer, i
 
             writer.Commit();
 
-            _unflushedBytes += writer.BytesCommitted;
             _autoChunk = autoChunk;
+            WriteDataWrittenBeforeHeaders(ref writer);
+
+            _unflushedBytes += writer.BytesCommitted;
+            _startCalled = true;
+        }
+
+        private void WriteDataWrittenBeforeHeaders(ref BufferWriter<PipeWriter> writer)
+        {
+            if (_completedSegments != null)
+            {
+                foreach (var segment in _completedSegments)
+                {
+                    if (_autoChunk)
+                    {
+                        CommitChunkInternal(ref writer, segment.Span);
+                    }
+                    else
+                    {
+                        writer.Write(segment.Span);
+                        writer.Commit();
+                    }
+                    segment.Return();
+                }
+
+                _completedSegments.Clear();
+            }
+
+            if (!_currentSegment.IsEmpty)
+            {
+                var segment = _currentSegment.Slice(0, _position);
+
+                if (_autoChunk)
+                {
+                    CommitChunkInternal(ref writer, segment.Span);
+                }
+                else
+                {
+                    writer.Write(segment.Span);
+                    writer.Commit();
+                }
+
+                _position = 0;
+
+                if (_currentSegmentOwner != null)
+                {
+                    _currentSegmentOwner.Dispose();
+                    _currentSegmentOwner = null;
+                }
```

Comment:
Should we also be setting _currentSegment to default?

Reply:
It will never be accessed again, so we don't need to. But I'll go ahead and set it.